Magic code blur

wojtus7 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

This PR is a direct continuation of the [PR](https://github.com/Expensify/App/pull/24012) that got reverted [here](https://github.com/Expensify/App/pull/28611).

It should his two release blockers:
https://github.com/Expensify/App/issues/28494
https://github.com/Expensify/App/issues/28590

Additionally resolve issue: 
https://github.com/Expensify/App/issues/26090

Description:
In order to make blur possible I needed to refactor the magic component. Instead of 6 separate text inputs there is only one wrapped inside TapGestureHandler that determine which number should be inputed. This resulted is some code that is simpler, taps are handled by gesture handler and right-click/context menu/context is handled by native text input. but at the same time created small issue that paste button shows only on the beginning of the input.

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/18244
PROPOSAL: https://github.com/Expensify/App/issues/18244#issuecomment-1642739632


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Log out
2. Provide email address
3. Try to break the magic code input (Click at random places, use arrows, backspace, change tabs and go back on web etc)

- [ ] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Log out
2. Provide email address
3. Try to break the magic code input (Click at random places, use arrows, backspace, change tabs and go back on web etc)

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/42337257/8b82c949-3449-405f-8a4d-bfbe761be886


</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/42337257/83908f32-ff5e-496d-97da-5a6dc8babbba

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/42337257/0483bcf6-be20-4298-bd1c-14f982ad067b

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/42337257/6a074fc4-9037-4354-a5eb-a2616784a126

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/42337257/c3123b29-3a92-4a13-911b-d6ffb93fbcbb

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/42337257/1ae65db0-6773-47ee-8bd5-773aaf060d52

</details>

wojtus7 committed:
Fix stuff

wojtus7 committed:
Merge remote-tracking branch 'origin/main' into magic-code-blur

wojtus7 commented:
@dangrous @Santhosh-Sellavel

wojtus7 committed:
Prettier

Santhosh-Sellavel commented:
@wojtus7 What's the new fix for the two blockers, can you explain for me please!

wojtus7 commented:
@Santhosh-Sellavel 

Issue https://github.com/Expensify/App/issues/28494:

So the problem here is that the code you get from SMS is not pasted but very rapidly inputed on the keyboard. In this implementation of magic input I get the value and clear the input after user type anything. If user paste anything I get 6 digits and use it and then clear input. The way iOS rapidly input new digits from SMS is breaking because it inputs the values so fast that react doesn't have time to clear the input because useState is async and takes a little time. So for code: 123456 the input in magic code was: 112123 because every change of the input was 1 -> 12 -> 123 -> 1234 and every state of the input was interpreted as pasted code.

The answer was to create new ref `const lastValue = useRef();` that will hold last input value and remove the shared beginning. If the code would be 123456 then for `1` -> `12` it would recognise the clear of the input did not happen and remove it from the string and just use `2`. 

At the same time we can't just not clear the input because the `paste` button on mobile devices will show in random places on the input not just on the first space :( 


Answer for https://github.com/Expensify/App/issues/28590 was just add `focus()` after modal closure

Santhosh-Sellavel commented:
@wojtus7 Sorry for the delay, please resolve conflicts!

Santhosh-Sellavel commented:
bump @wojtus7

dangrous commented:
hi, just letting y'all know I'm back from OOO and happy to review this whenever!

dangrous commented:
@wojtus7 I'd love to get this merged this week - do you have time to resolve conflicts and get this ready to review again? Thank you!

wojtus7 commented:
Hi @dangrous Today I'm first day back after being OOO for 2 weeks. I started to work on it right away and have all conflicts resolved but my GPG stopped working after not using Mac for some reason and trying to figure it out. Working on it right now, will push the changes whenever it's resolved and will do my best to make this PR merge this week!

wojtus7 committed:
Merge remote-tracking branch 'origin/main' into magic-code-blur

wojtus7 commented:
@Santhosh-Sellavel all conflicts resolved!

dangrous commented:
Thank you @wojtus7 ! @Santhosh-Sellavel let us knwo if you have time to review today or tomorrow - thank you!

Santhosh-Sellavel commented:
TestFails on Android.

The keyboard disappears immediately when magic code input is presented

https://github.com/Expensify/App/assets/85645967/41f85e24-aeaa-4f4e-a149-244d82b5a6cb

Santhosh-Sellavel commented:
This https://github.com/Expensify/App/issues/28590 is not resolved 


https://github.com/Expensify/App/assets/85645967/caaab53a-094d-4d67-99c9-8e019cd8ef3a

Santhosh-Sellavel commented:
@wojtus7 
https://github.com/Expensify/App/issues/26090 This is not resolved either 

https://github.com/Expensify/App/assets/85645967/81117dbf-d5b7-480f-8461-0f2f6a60aa49


### 🔴 🔴 While testing found even a weirder case was able to enter more than 6 numbers but it changed only the last digit and the code was not submitted at all.

https://github.com/Expensify/App/assets/85645967/5d45f753-cb11-4516-8fb8-102cdc9f3197

cc: @dangrous

wojtus7 commented:
Hmm. I guess some things break after the main merge or I am going crazy 😞
I'm looking at thise things right now

@Santhosh-Sellavel about:
"🔴 🔴 While testing found even a weirder case was able to enter more than 6 numbers but it changed only the last digit and the code was not submitted at all."

I think this is caused by and the auto submit feature being turn off after one time:
https://github.com/Expensify/App/pull/29731

wojtus7 committed:
Merge remote-tracking branch 'origin/main' into magic-code-blur

wojtus7 committed:
Refocus input in basevalidatecodeform and remove wasSubmitted flag

wojtus7 commented:
> TestFails on Android.
> 
> The keyboard disappears immediately when magic code input is presented
> 
>  Screen.Recording.2023-10-26.at.12.08.38.AM.mov

I think this is more complex than that. I tested it a lot and it happen in a lot of places and not only in the way you described but also can happen when go back a screen then the email input fails, and sometimes the error doesn't happen at all. 
Trying to fix it locally in magic code but maybe this is some navigation issue?

https://github.com/Expensify/App/assets/42337257/ea749006-b324-44df-b203-2e45aacd2387

wojtus7 committed:
Use InteractionManager instead of delay

wojtus7 committed:
Prettier

Santhosh-Sellavel commented:
Can you resolve conflicts, please? Whats the status here?

wojtus7 committed:
Merge remote-tracking branch 'origin/main' into magic-code-blur

wojtus7 committed:
Merge remote-tracking branch 'origin/main' into magic-code-blur

wojtus7 committed:
Use @ imports

wojtus7 committed:
Improve import

wojtus7 committed:
Prettier

wojtus7 commented:
Status here is:
* I removed the `wasSubmitted` logic so the _"🔴 🔴 While testing found even a weirder case was able to enter more than 6 numbers but it changed only the last digit and the code was not submitted at all."_ doesn't happen anymore
* Fixed refocus issue https://github.com/Expensify/App/issues/28590
* Looking into https://github.com/Expensify/App/issues/26090

dangrous commented:
@Santhosh-Sellavel I know you were out for a little bit, would you prefer us reassign here?

Santhosh-Sellavel commented:
@dangrous I'm back now, waiting for the below from this [comment](https://github.com/Expensify/App/pull/28711#issuecomment-1785193951)

> Looking into https://github.com/Expensify/App/issues/26090

dangrous commented:
@wojtus7 let us know what you find out, and if you need any help!

wojtus7 commented:
@Santhosh-Sellavel @dangrous 
I tried to fix the https://github.com/Expensify/App/issues/26090 but without great success and I don't think it's easy to resolve it.

Since the reason for revert was https://github.com/Expensify/App/issues/28494 and https://github.com/Expensify/App/issues/28590 which are fixed now (I think) and the https://github.com/Expensify/App/issues/26090 is reproducable on both main and this branch do you think it is reasonable to consider it separate bug that is not a blocker for this PR?

dangrous commented:
I think I'm okay with that - it's probably good to get some other eyes on #26090 and not have this hold that up. So if this fixes the necessary blockers, than I'm okay with it! Does that sound alright to you, @Santhosh-Sellavel ?

Santhosh-Sellavel commented:
We can leave https://github.com/Expensify/App/issues/26090. 
As we are doing a refactor for holistically addressing inconsistencies & bugs, so I'm wondering whether it might end up as refactor.

We can give another try to address it here. Mean while I'll go through tests & checklist again.

cc: @dangrous @wojtus7

dangrous commented:
Yeah, I think it makes sense for now. Let us know how the tests/checklist goes!

Santhosh-Sellavel commented:
@wojtus7 Can you resolve conflicts please

wojtus7 committed:
Merge remote-tracking branch 'origin/main' into magic-code-blur

Santhosh-Sellavel commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

Santhosh-Sellavel commented:
Bug: Something else is broken after entering the wrong code. Unable to enter/paste the right code without clearing it. 

https://github.com/Expensify/App/assets/85645967/f9b4b658-a86b-4d2d-8188-21e9ac5ccc50

The code is getting submitted after entering each digit.

https://github.com/Expensify/App/assets/85645967/d8f05e3c-8258-404b-b8ce-b145c8e1e34c

Santhosh-Sellavel commented on src/components/MagicCodeInput.js:
```diff
@@ -285,6 +349,48 @@ function MagicCodeInput(props) {
+                            }}
+                            selectionColor="transparent"
+                            inputStyle={[styles.inputTransparent]}
+                            accessibilityRole={CONST.ACCESSIBILITY_ROLE.TEXT}
```
getting a warning for this
<img width="357" alt="Screenshot 2023-11-08 at 11 28 04 PM" src="https://github.com/Expensify/App/assets/85645967/e4607bd9-e574-4c89-b562-a73075eca529">

Santhosh-Sellavel commented:
Regarding https://github.com/Expensify/App/pull/28711#issuecomment-1802382433 
This is how it works on staging, i.e. after the wrong code error only when the user manually submits code gets submitted. I'm confused about whether the behavior on staging is correct or not. @dangrous can you confirm?

https://github.com/Expensify/App/assets/85645967/98c60e58-aad4-4044-b10d-56197ed8f9e0

dangrous commented:
@Santhosh-Sellavel Yep, I agree with that. I'd like this to get settled.

I think staging behavior seems correct to me - after a wrong code, only a user actually clicking the button should submit. But yeah those other two bugs you shared def should be fixed.

wojtus7 committed:
Merge remote-tracking branch 'origin/main' into magic-code-blur

wojtus7 committed:
Use role instead of accessibilityRole

wojtus7 committed:
Resurrect wasSubmitted flag

wojtus7 committed:
Add additional checks for code input paste and fast typing

wojtus7 committed:
Fix lint

kosmydel committed:
add shouldDelayFocus

kosmydel commented:
Hey @Santhosh-Sellavel, can I ask for another round of review? 
@wojtus7 addressed the bugs and I've updated the PR to cover also [this issue](https://github.com/Expensify/App/issues/30281)

Santhosh-Sellavel commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

Santhosh-Sellavel commented:
The bugs mentioned [here](https://github.com/Expensify/App/pull/28711#issuecomment-1802382433) are fixed now. So far everything works well. Will continue testing tomorrow

Santhosh-Sellavel commented:
@wojtus7 @dangrous @kosmydel 

Mostly looks good. While testing today I ran into this issue while testing on Android Alone.

Tested this issue https://github.com/Expensify/App/issues/28590. Focus is set correctly but the keyboard is not shown on the Android

https://github.com/Expensify/App/assets/85645967/e8fee735-f24b-403b-a30f-297874bdb2d1

This is NAB as it is specific to a platform and can be handled as a follow-up. 

Basic cases pass for me, so I'm good with approving this. But I couldn't test this [issue](https://github.com/Expensify/App/issues/28494) on iOS Simulator can you test this please @dangrous iPhone, thanks!

dangrous commented:
Oh hm I'm not sure I can test that even with a physical iPhone since I need it to text me and I don't think it will... Let me investigate.

dangrous commented:
In the meantime, if we CAN find a fix for the keyboard thing you mentioned that would be ideal, but yeah I don't think it's a blocker

kosmydel committed:
fix android keyboard

kosmydel committed:
Merge branch 'main' into @kosmydel/magic-code-blur

kosmydel committed:
Refactor focusTimeoutRef.current in
BaseValidateCodeForm.js

kosmydel commented:
> In the meantime, if we CAN find a fix for the keyboard thing you mentioned that would be ideal, but yeah I don't think it's a blocker

I've fixed this issue [here](https://github.com/Expensify/App/pull/28711/commits/92f5be143234fdbabcf5f6991d1c119f6a698e0e), pulled main & resolved conflicts. 

**Video:**

https://github.com/Expensify/App/assets/104823336/78860430-941a-4d41-9892-fb1764b1009c

I've tested the magic code, but I couldn't test the OTP auto-fill functionality (I can't send text messages to the testing devices that I have).

cc @Santhosh-Sellavel for another review

dangrous commented:
Okay I managed to set up getting texts on my physical iOS device.

BUT - both on staging, main, and this branch - the phone does not try to autofill from a text anymore - I don't get the prompt. Can one of you possibly try on **staging** and see if it's my phone or something else? It does look like the messaging changed since that bug was filed, so that could be it.

If you can't, I can ask around internally, too. Thanks!

Santhosh-Sellavel commented:
@dangrous  I can't test this case alone, please ask internally.

dangrous commented:
Okay, I figured it out! Seems to work well.

kosmydel commented:
Hey, when can I expect another round of review? @Santhosh-Sellavel @dangrous

dangrous commented:
@Santhosh-Sellavel are you able to test/review again? The iOS issue seems to be solved with this PR.

Santhosh-Sellavel commented:
AFK today, will do tomorrow.

Santhosh-Sellavel commented:
On it now again!

Santhosh-Sellavel reviewed (approved):
LGTM, tests well!

dangrous commented on src/components/MagicCodeInput.js:
```diff
@@ -244,24 +301,37 @@ function MagicCodeInput(props) {
+            // Saves the input string so that it can compare to the change text
+            // event that will be triggered, this is a workaround for mobile that
+            // triggers the change text on the event after the key press.
+            setInput(TEXT_INPUT_EMPTY_STATE);
+            setFocusedIndex(newFocusedIndex);
+            setEditIndex(newFocusedIndex);
```
We use these three lines in a number of places - might it be good to separate them out into a helper function?

dangrous commented on src/components/MagicCodeInput.js:
```diff
@@ -181,9 +233,12 @@ function MagicCodeInput(props) {
             return;
         }
 
+        const addedValue =
```
Can you add a comment to explain this? It's very confusing to look at, haha. Also some parentheses around the boolean section might help.

dangrous reviewed (changes requested):
Two minor clarifications, otherwise this is looking good!

kosmydel committed:
add helper function

kosmydel committed:
add comments

dangrous reviewed (approved):
love it!

dangrous commented:
Okay, going to merge, let's be sure to watch this for bugs but I think we're going to be good this time.

dangrous commented on src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:
```diff
@@ -264,6 +264,11 @@ class ContactMethodDetailsPage extends Component {
                         onCancel={() => this.toggleDeleteModal(false)}
+                        onModalHide={() => {
+                            InteractionManager.runAfterInteractions(() => {
+                                this.validateCodeFormRef.current.focusLastSelected();
```
I can handle the fix if need be but if you're around @wojtus7 that would be great!

kosmydel replied:
Hey, @wojtus7 is not here. I can help with this one in an hour or two.

ZhenjaHorbach replied:
I can help too )

kosmydel replied:
Working on the PR

dangrous replied:
Thank you for jumping in @kosmydel! We'll be ready to review whenever

kosmydel replied:
[PR](https://github.com/Expensify/App/pull/32193)

I'm recording videos, however, I have a rate limit for adding new contact methods.

kosmydel replied:
[The PR](https://github.com/Expensify/App/pull/32193) is ready :)
cc @situchan @dangrous

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7046178910) to production by https://github.com/mountiny in version: 1.4.5-7 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

alitoshmatov commented on src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.js:
```diff
@@ -89,13 +89,25 @@ function BaseValidateCodeForm(props) {
+            if (focusTimeoutRef.current) {
+                clearTimeout(focusTimeoutRef.current);
+            }
             focusTimeoutRef.current = setTimeout(inputValidateCodeRef.current.focus, CONST.ANIMATED_TRANSITION);
```
`focusLastSelected` should have been used also here, like this:
```  focusTimeoutRef.current = setTimeout(inputValidateCodeRef.current.focusLastSelected, CONST.ANIMATED_TRANSITION);`

This caused issue - https://github.com/Expensify/App/issues/33170
